### ISSUE TITLE
Docker proxy using ports after removal of containers.

### ISSUE THREAD

**Issue by bweston92** [docker/docker#26527]
**Description**

```
[root@localhost bweston]# docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
[root@localhost bweston]# netstat --tcp --udp --listening --program | grep 15672
tcp6       0      0 [::]:15672              [::]:*                  LISTEN      23076/docker-proxy  
[root@localhost bweston]# systemctl stop docker
[root@localhost bweston]# netstat --tcp --udp --listening --program | grep 15672
[root@localhost bweston]# systemctl start docker
[root@localhost bweston]# netstat --tcp --udp --listening --program | grep 15672
tcp6       0      0 [::]:15672              [::]:*                  LISTEN      10205/docker-proxy  
[root@localhost bweston]# docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES

# Creating a new container with the port gives me
# ERROR: driver failed programming external connectivity on endpoint rabbitmq_1 (9255fe0bee85c10f2508ca336469145d4121b78c8bd7594d6c5bc11814bb0eab): Bind for 0.0.0.0:15672 failed: port is already allocated
```

**Steps to reproduce the issue:**
1. Create container with port exposed.
2. Remove container.
3. Create container with port exposed and you get an error.

**Describe the results you received:**

Unable to start a new container with port after old container removed.

**Describe the results you expected:**

Being able to start a new container with port after old container removed.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Docker version 1.12.1, build 23cf638
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 226
Server Version: 1.12.1
Storage Driver: devicemapper
 Pool Name: docker-8:3-2621441-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 18.58 GB
 Data Space Total: 107.4 GB
 Data Space Available: 88.79 GB
 Metadata Space Used: 30.77 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.117 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.109 (2015-09-22)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.7.2-101.fc23.x86_64
Operating System: Fedora 23 (Workstation Edition)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 11.53 GiB
Name: localhost.localdomain
ID: XRPK:74WN:CHAZ:RR3A:JS3J:Q7G5:EXHM:7OCD:IB5J:SMFF:YZ27:T52Q
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: bweston92
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Physical

**Comment by aboch** [comment]
@bweston92 Thanks for the report.
Can you please provide more information about the reproduction steps, how the container is started.
For example I am assuming the container has a restart policy, what is the restart policy ?

**Comment by thaJeztah** [comment]
ping @bweston92

**Comment by ctammes** [comment]
Same problem here.

Last Tuesday I had this errror. I removed all container and restarted the daemon and the problem was over.
But now it happens again. It effects two of the four running containers, which are all in the same network and connected to each other. It happens with ports 3000, 3001 and 3316.

I did following to "solve":

1. stop all running containers: **docker stop <container>**
2. remove all containers: **docker rm $(docker ps -aq)**
3. remove network: **docker network rm <network>**
4. stop daemon: **sudo service docker stop**
5. verify no ports are in use: **netstat -an | egrep ":(3000|3001|3316)"**
6. start daemon: **sudo service docker start**
7. verify no ports are in use: **netstat -an | egrep ":(3000|3001|3316)"**
8. create network: **docker network create <network>**
9. build/start containers: **docker run --restart=always ...**

-p usage:
-p 3316:3306
-p 0.0.0.0:3000:3000 -p 0.0.0.0:3001:3001

Docker version:
```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:22:43 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:22:43 2016
 OS/Arch:      linux/amd64
```

uname -a:
`Linux HP-Compaq-dc7900-Convertible-Minitower 3.13.0-96-generic #143-Ubuntu SMP Mon Aug 29 20:15:20 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by thaJeztah** [comment]
@ctammes do you have swarm mode enabled? @aboch could it be related to https://github.com/docker/docker/issues/26404 in that case?

**Comment by aboch** [comment]
@thaJeztah I do not think so, given in @bweston92 's docker info o/p swarm is inactive, and @ctammes is explicitly starting the containers with `docker run`.

**Comment by ctammes** [comment]
@thaJeztah No, I don't have swarm mode.

**Comment by aboch** [comment]
Looking better at this, @bweston92's issue is a duplicate of #25981.
See https://github.com/docker/docker/issues/25981#issuecomment-254663996 for the details.

**Comment by mavenugo** [comment]
Thanks @aboch. yes. I agree. closing this. feel free to reopen it if this is not resolved after #25981 is closed.
